### PR TITLE
aggressively flush cache based on cache size

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Summary of Changes

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
Here is a time when cache size is growing large, but we are below 200 slots:
image

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -4515,6 +4524,14 @@ impl AccountsDb {
         self.flush_slot_cache(slot, None::<&mut fn(&_, &_) -> bool>);
     }
 
+    /// true if write cache is too big or has too many slots
+    fn should_aggressively_flush_cache(&self) -> bool {
+        self.write_cache_limit_bytes
+            .map(|write_cache_limit_bytes| self.accounts_cache.size() >= write_cache_limit_bytes)
+            .unwrap_or_default()
+            || self.accounts_cache.num_slots() > MAX_CACHE_SLOTS
```
Comment (carllin) [inline runtime/src/accounts_db.rs:4532]: If we're going with the size based approach, we can probably get rid of the `MAX_CACHE_SLOTS` check

File: runtime/src/accounts_cache.rs [hunk 2]
```diff
@@ -289,26 +289,21 @@ impl AccountsCache {
...
-                    }
-                })
-                .collect();
-            slots.sort_unstable();
-            slots.truncate(slots.len().saturating_sub(num_to_retain));
-            slots
-        } else {
-            vec![]
-        }
+    pub fn find_older_frozen_slots(&self) -> Vec<Slot> {
```
Comment (carllin) [inline runtime/src/accounts_cache.rs:292]: nit: this just returns all frozen slots now, so would rename to `cached_frozen_slots`

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -4560,26 +4577,34 @@ impl AccountsDb {
...
+                    } else {
+                        unflushable_unrooted_slot_count += 1;
+                        None
+                    }
+                })
+                .collect();
+            info!(
+                "req_flush_root: {:?} old_slot_flushes: {:?}",
+                requested_flush_root, old_slot_flush_stats
+            );
```
Comment (carllin) [inline runtime/src/accounts_db.rs:4606]: Might make sense to turn this into a datapoint capturing:

1. Number of slots in the cache at the time the aggressive flush happened
2. An aggregate of the measures in `old_slot_flush_stats`

File: runtime/src/accounts_db.rs [hunk 4]
```diff
@@ -4560,26 +4577,34 @@ impl AccountsDb {
...
-        );
+        if self.should_aggressively_flush_cache() {
+            let old_slots = self.accounts_cache.find_older_frozen_slots();
+            excess_slot_count = old_slots.len();
+            let old_slot_flush_stats: Vec<_> = old_slots
+                .into_iter()
+                .filter_map(|old_slot| {
+                    // Don't flush slots that are known to be unrooted
+                    if old_slot > max_flushed_root {
+                        if self.should_aggressively_flush_cache() {
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:4592]: We check here again because the cache size could be reduced after flushing, is that right?

Comment (jeffwashington) [reply runtime/src/accounts_db.rs:4592]: yes, each successful call to flush_slot_cache could cause us to drop below the limit(s). Previously, we were only aggressively flushing on # slots, so we knew how many slots to aggressively flush.

**Comment by sakridge** [comment]
lgtm, +1 to what @carllin said

File: runtime/src/accounts_db.rs (hunk 3 above)
Comment (jeffwashington) [reply runtime/src/accounts_db.rs]: aggregated.

File: runtime/src/accounts_db.rs [hunk 5]
```diff
@@ -4566,26 +4580,40 @@ impl AccountsDb {
...
-            "req_flush_root: {:?} old_slot_flushes: {:?}",
-            requested_flush_root, old_slot_flush_stats
-        );
+            });
+            datapoint_info!(
+                "accounts_db-flush_accounts_cache_aggressively",
+                ("num_flushed", flush_stats.num_flushed, i64),
+                ("num_purged", flush_stats.num_purged, i64),
+                ("total_size", flush_stats.total_size, i64),
+                ("total_size", flush_stats.total_size, i64),
```
Comment (carllin) [inline runtime/src/accounts_db.rs:4612]: duplicate of line above, should this be the total size of the cache currently?

**Review by carllin — APPROVED** [review]

**Pull request merged**